lib/utils.nix

```diff
@@ -125,27 +125,185 @@ rec {
     end
   '';
 
+  /**
+    Generic "heredoc-style" escape utility.
+
+    # Args
+
+    - string: The string to escape.
+    - buildSurround: A function that maps Integer to Surround. Described in detail below.
+    - surroundWithNewlines: A boolean that impacts how to surround the string when escaping it.
+
+    A Surround is an attr set that looks like this:
+
+    ```nix
+    {
+      open = "(";
+      close = ")";
+    }
+    ```
+
+    `buildSurround` is a function that takes a depth Integer and returns a
+    Surround. For exmple, to build the previous Surround:
+
+    ```nix
+    nix-repl> buildSurround = depth: {
+      open = lib.strings.replicate depth "(";
+      close = lib.strings.replicate depth ")";
+    }
+    nix-repl> buildSurround 1
+    {
+      close = ")";
+      open = "(";
+    }
+    ```
+
+    We invoke `buildSurround` builder starting a depth 0, and increasing the
+    depth until we have a Surround where `surround.close` is not in the
+    string we're escaping.
+
+    We return the original string, first (possibly, depending on
+    `surroundWithNewlines`) surrounded by newlines , and secondly surrounded by
+    the Surround we found above.
+    ```
+  */
+  escapeString =
+    {
+      string,
+      buildSurround,
+      surroundWithNewlines,
+    }:
+    let
+      findSurround =
+        {
+          string,
+          buildSurround,
+          depth,
+        }:
+        let
+          surround = buildSurround depth;
+        in
+        if lib.hasInfix surround.close string then
+          findSurround {
+            inherit string buildSurround;
+            depth = depth + 1;
+          }
+        else
+          surround;
+
+      surround = findSurround {
+        inherit string buildSurround;
+        depth = 0;
+      };
+      newlines = if surroundWithNewlines then "\n" else "";
+    in
+    "${surround.open}${newlines}${string}${newlines}${surround.close}";
+
+  /**
+    Given a string, escape it as a Lua [long literal].
```

Comment:
I don't think this really relates to escaping, so I'm interested in alternate naming...

To me, escaping might be converting `\n` into `\\n`, for example.

I think what we're doing here is safely _embedding_ a string into some language using enclosing syntax that we are confident wraps the string without an early termination.

I guess, conceptually, it is similar to escaping (otherwise) terminating quotes from a string literal. However functionally we implement it differently; we aren't escaping the string, we are adapting the opening/closing syntax to better fit the string.

Reply:
I'm just realizing I invented the noun "Surround" and the verb "surround" for this PR when there was already the word "wrap" in this repo (and I could have invented the noun "Wrap"). I'm not sure there's much value in having both terms floating around.

Let me know what you think is best, @MattSturgeon

Reply:
Hm. Naming is always a challenge!

Personally I like the names that are clear about the end result, so `toLuaLongLiteral` and `toVimscriptHeredoc` would be my current preference. Good suggestions!

We could be even more explicit and replace `to` with `wrapAs`, but I think that's unnecessary.